### PR TITLE
Refactor: Rewrite loaders folder in typescript

### PR DESCRIPTION
**Pull request by elevatebart**

### CODE REVIEW

**Comment by elevatebart** [comment]
Just a heads up, this might actually be a breaking change for those who have written typings for their rsg based project. 

Should I remove the publishing of typings?

**Comment by elevatebart** [comment]
To avoid any potential breaking changes, I postponed the publishing of the types until we have proper testing and documentation for it.

**No more potential breaking change**

File: src/bin/__tests__/styleguidist.spec.ts [hunk 1]
```diff
@@ -0,0 +1,13 @@
+import fs from 'fs';
+import vm from 'vm';
+import path from 'path';
+import stripShebang from 'strip-shebang';
+
+const BIN_PATH = path.resolve(__dirname, '../styleguidist.js');
+
+// Very simple test to be sure we’re not using any syntax that is not supported in versions of Node we support
+it.skip('should not throw when parsing a script', () => {
```
Comment (sapegin) [inline src/bin/__tests__/styleguidist.spec.ts]: 🤔

File: src/loaders/utils/getComponentPatternsFromSections.ts [hunk 2]
```diff
@@ -1,12 +1,15 @@
...
+ * NOTE: a section cannot have components & subsections
  * @param {Array} sections
  * @returns {Array}
  */
-module.exports = function getComponentPatternsFromSections(sections) {
-	return sections.reduce((patterns, section) => {
-		if (section.components) {
+export default function getComponentPatternsFromSections(sections: Rsg.ConfigSection[]): string[] {
+	return sections.reduce((patterns: string[], section) => {
+		if (isArray(section.components)) {
```
Comment (sapegin) [inline src/loaders/utils/getComponentPatternsFromSections.ts]:
```suggestion
		if (Array.isArray(section.components)) {
```

File: src/bin/__tests__/styleguidist.spec.ts (hunk 1 above)
Comment (elevatebart) [reply src/bin/__tests__/styleguidist.spec.ts]: yes, this file should not exist. I believe we should not transpile this `bin` folder.

**Comment by elevatebart** [comment]
Now I am good with the PR. Merge when you want.

**Pull request merged**